profitbricks/src/main/java/org/jclouds/profitbricks/domain/Firewall.java

```diff
@@ -0,0 +1,367 @@
+	    private List<RuleWithIcmp> rules = Lists.newArrayList();
+
+	    public Builder nicId(String nicId) {
+	       this.nicId = nicId;
+	       return this;
+	    }
+
+	    public Builder rules(List<RuleWithIcmp> rules) {
+	       this.rules = rules;
+	       return this;
+	    }
+
+	    public RuleWithIcmp.Builder newRule() {
+	       return new RuleWithIcmp.Builder(this);
+	    }
+
+	    public Builder addRule(RuleWithIcmp rule) {
+	       this.rules.add(rule);
+	       return this;
+	    }
+
+	    public AddRulePayload build() {
+	       return AddRulePayload.create(nicId, rules);
+	    }
+	 }
+      }
+   }
+
+   public static class Builder {
+
+      private String id;
+      private String nicId;
+      private boolean active;
+
+      private ProvisioningState state;
+      private List<Rule> rules;
+
+      public Builder id(String id) {
+	 this.id = id;
+	 return this;
+      }
+
+      public Builder nicId(String nicId) {
+	 this.nicId = nicId;
+	 return this;
+      }
+
+      public Builder active(boolean active) {
+	 this.active = active;
+	 return this;
+      }
+
+      public Builder state(ProvisioningState state) {
+	 this.state = state;
+	 return this;
+      }
+
+      public Builder rules(List<Rule> firewallRules) {
+	 this.rules = firewallRules;
+	 return this;
+      }
+
+      public Builder fromFirewall(Firewall in) {
+	 return this.id(in.id()).nicId(in.nicId()).active(in.active()).state(in.state())
+		 .rules(in.rules());
+      }
+
+      public Firewall build() {
+	 return Firewall.create(id, nicId, active, state, rules);
+      }
+   }
+
+   public abstract static class RuleBuilder<B extends RuleBuilder, D extends FirewallRuleCommonProperties> {
+
+      protected String name;
+      protected Integer portRangeEnd;
+      protected Integer portRangeStart;
+      protected Protocol protocol;
+      protected String sourceIp;
+      protected String sourceMac;
+      protected String targetIp;
+
+      public B name(String name) {
+	 this.name = name;
+	 return self();
+      }
+
+      public B portRangeEnd(Integer portRangeEnd) {
+	 this.portRangeEnd = portRangeEnd;
+	 return self();
+      }
+
+      public B portRangeStart(Integer portRangeStart) {
+	 this.portRangeStart = portRangeStart;
+	 return self();
+      }
+
+      public B protocol(Protocol protocol) {
+	 this.protocol = protocol;
+	 return self();
+      }
+
+      public B sourceIp(String sourceIp) {
+	 this.sourceIp = sourceIp;
+	 return self();
+      }
+
+      public B sourceMac(String sourceMac) {
+	 this.sourceMac = sourceMac;
+	 return self();
+      }
+
+      public B targetIp(String targetIp) {
+	 this.targetIp = targetIp;
+	 return self();
+      }
+
+      public abstract B self();
+
+      public abstract D build();
+
+      protected void checkPortRange() {
+	 checkArgument((portRangeEnd == null && portRangeStart == null)
+		 || (portRangeEnd != null && portRangeStart != null), "Port range must be both present or null");

```

Comment:
[nit] This can be more concise:
```java
checkArgument(!(portRangeEnd == null ^ portRangeStart == null), "Port range must be both present or null");
```